### PR TITLE
20230201 defconst

### PR DESCRIPTION
**Pull request by prozacchiwawa**
Adds a defconst form whose body is interpreted as chialisp and which has access to the functions defined in the parent program.
Mostly useful for calling sha256tree on constant values to pre-interpret them into compile time constants.
Constants can use other constants, but dependency loops aren't allowed.

            (mod ()
              (include sha256tree.clib)
              (defconst H (+ G (sha256tree (q 2 3 4))))
              (defconst G 1)
              H
              )

Because 

    $ opc -H '(2 3 4)'
    6fcb06b1fe29d132bb37f3a21b86d7cf03d636bf6230aa206486bef5e68f9873

this program compiles to

    (q . 0x6fcb06b1fe29d132bb37f3a21b86d7cf03d636bf6230aa206486bef5e68f9874)

### CODE REVIEW

**Comment by arvidn** [comment]
I was under the impression that the compiler (generally) is very aggressive in constant-folding, and that calling `sha256tree` on anything that's a constant expression will be replaced by the result. I can see an advantage of `defconst` in that (presumably) it's a compilation error if the expression is not constant. Perhaps it also makes it easier to reuse the same atom in the final compiler code, rather then encoding the same constant multiple times. however, I would expect the latter could also be solved with string-literal de-duplication (for atoms)

**Comment by prozacchiwawa** [comment]
It is but there are a couple reasons defconst is better.

defconstant in classic chialisp does no evaluation on its argument (treated literally as quoted), so if you want an expression involving any operations (even (+ 1 2)), you have to write it in the code on top of a constant that can be expressed literally and the optimizer will indeed fold it.  However, the constant folding in classic chialisp is _very_ slow for tree forms.  If you recall, I landed a PR a while ago that added caching to classic chialisp compilation.  That improved runtime of a certain constant treehash from over 6 hours to roughly 6 minutes (on my core i5), which still isn't ideal by any stretch but it became liveable.  I haven't characterized the classic optimizer (imo) well enough to tear into it and make it much faster again, and might not be able to provably do so... for example, var_change_optimizer_cons_eval applies in pre-order, which means that the form of the expression can change many times (and be re-optimized many times) before reaching all leaves.

It's also the case that defconstant doesn't have access to the program's functions or other constants.  People have used macros instead (and individually re-included the program's library functions to compute constants in macros, but that's also gross).  defconst simply composes the program you'd want to generate your constant value and runs it.

defconst in both languages is (IMO) a more sensible form.  It interprets its argument in the same way as other expressions in the language and like (define ) in scheme or (setq ) in lisp, has access to (conceptually) the program that's been entered before it.

So that's why a separate form.  Hoping defconst can just catch on everywhere.

File: src/classic/clvm_tools/stages/stage_2/module.rs [hunk 1]
```diff
@@ -327,30 +336,30 @@ fn parse_mod_sexp(
...
+                let r_of_declaration = rest(allocator, declaration_sexp)?;
+                let rr_of_declaration = rest(allocator, r_of_declaration)?;
+                let frr_of_declaration = first(allocator, rr_of_declaration)?;
+                let quoted_decl = quote(allocator, frr_of_declaration)?;
+                constants.insert(name, quoted_decl);
+                Ok(())
+            } else if op == "defconst".as_bytes() {
+                let r_of_declaration = rest(allocator, declaration_sexp)?;
+                let rr_of_declaration = rest(allocator, r_of_declaration)?;
+                let frr_of_declaration = first(allocator, rr_of_declaration)?;
```
Comment (arvidn) [inline src/classic/clvm_tools/stages/stage_2/module.rs]: this seems a bit unnecessarily verbose. compared to just
```
let frr_of_declaration = first(allocator, rest(allocator, rest(allocator, declaration_sexp)?)?)?;
```

File: src/classic/clvm_tools/stages/stage_2/module.rs [hunk 2]
```diff
@@ -385,26 +397,110 @@ fn compile_mod_stage_1(
...
+                    macros
+                };
+
+                let main_name_vec = MAIN_NAME.as_bytes().to_vec();
+
+                // Process delayed constants until we either can't advance or
+                // they're all done.
+
+                loop {
+                    if delayed_constants.is_empty() {
```
Comment (arvidn) [inline src/classic/clvm_tools/stages/stage_2/module.rs:422]: it looks like this `loop` could just be a for-loop. The only reason you need to mutate this vector (as far as I can tell) is because you delete the entries after you've processed it. But I don't think you really need to remove them, do you?
There's so much mutable state here, it's a bit hard to follow along.

File: src/classic/clvm_tools/stages/stage_2/module.rs [hunk 3]
```diff
@@ -385,26 +397,110 @@ fn compile_mod_stage_1(
...
+                        break;
+                    }
+
+                    let mut processed = false;
+                    // copy so we can modify delayed_constants.
+                    let delayed_constant_defs: Vec<(Vec<u8>, NodePtr)> =
+                        delayed_constants.iter().map(|(k,v)| (k.clone(), *v)).collect();
+
+                    for (name, delayed_body) in delayed_constant_defs.iter() {
```
Comment (arvidn) [inline src/classic/clvm_tools/stages/stage_2/module.rs:431]: I believe `.iter()` is redundant here, isn't it?

File: src/classic/clvm_tools/stages/stage_2/module.rs (hunk 2 above)
Comment (arvidn) [reply src/classic/clvm_tools/stages/stage_2/module.rs:422]: I think I see why you do this. You don't know the order in which they need to be evaluated up-front, so you loop until they're all evaluated.

Do you have a test where you have cycles? And also other mixtures of orders.

File: src/classic/clvm_tools/stages/stage_2/module.rs [hunk 4]
```diff
@@ -385,26 +397,110 @@ fn compile_mod_stage_1(
...
-                        functions,
-                        constants,
-                        macros
-                    })
+                    if !processed {
+                        return Err(EvalErr(allocator.null(), "got stuck untangling defconst dependencies".to_string()));
+                    }
                 }
+
+                let uncompiled_main = alist[alist.len() - 1];
```
Comment (arvidn) [inline src/classic/clvm_tools/stages/stage_2/module.rs]: would `alist.last();` work here?

File: src/tests/classic/run.rs [hunk 5]
```diff
@@ -281,6 +281,107 @@ fn test_forms_of_destructuring_allowed_by_classic_1() {
...
+              (defconst G (logand H 1))
+              H
+              )
+        "}
+        .to_string(),
+    ])
+    .trim()
+    .to_string();
+    assert!(result_text.starts_with("FAIL"));
+    assert!(result_text.contains("got stuck untangling defconst dependencies"));
```
Comment (arvidn) [inline src/tests/classic/run.rs:403]: I see you do

File: src/classic/clvm_tools/stages/stage_2/module.rs (hunk 1 above)
Comment (arvidn) [reply src/classic/clvm_tools/stages/stage_2/module.rs]: oh, maybe the problem is the borrow-checker doesn't like `allocator` multiple times there, is that the case?

Comment (prozacchiwawa) [reply src/classic/clvm_tools/stages/stage_2/module.rs]: Yes, there aren't sequence points between the expressions when they're nested.  I was writing this in monadic style when i learned rust, but definitely should switch to ? style (as i have in other places).  This was made to be structured precisely identically to the code it was ported from but as I've gained confidence with it, some has got refactored, and this is likely overdue.

I've been considering match languages for a while and did some work on thinking about capturing values by position, so it's probably time to pop that and make this nicer.

File: src/classic/clvm_tools/stages/stage_2/module.rs (hunk 2 above)
Comment (prozacchiwawa) [reply src/classic/clvm_tools/stages/stage_2/module.rs:422]: The tests include cycles and out of order sequence but I'll add more.

File: src/classic/clvm_tools/stages/stage_2/module.rs (hunk 3 above)
Comment (prozacchiwawa) [reply src/classic/clvm_tools/stages/stage_2/module.rs:431]: So far clippy hasn't bugged me.  I'll check.

File: src/classic/clvm_tools/stages/stage_2/module.rs (hunk 4 above)
Comment (prozacchiwawa) [reply src/classic/clvm_tools/stages/stage_2/module.rs]: Probably.  I should get in that habit.  I need to come up with a good pattern for unexpected error in those cases, as exposing the Option gives the possibility of catching the err and one likely should.

File: src/classic/clvm_tools/stages/stage_2/module.rs (hunk 2 above)
Comment (prozacchiwawa) [reply src/classic/clvm_tools/stages/stage_2/module.rs:422]: I actually later made a general toposort algorithm and put it in util (second time writing it, for (assign ...) form in a different pr.  I considered backporting it here but thought it might be better to introduce it in a different pr.  If it seems like a good idea i'll pull that into a separate pr and remove the bespoke toposort here.

File: src/classic/clvm_tools/stages/stage_2/module.rs (hunk 1 above)
Comment (prozacchiwawa) [reply src/classic/clvm_tools/stages/stage_2/module.rs]: See what you think of that.

File: src/classic/clvm_tools/stages/stage_2/module.rs (hunk 2 above)
Comment (prozacchiwawa) [reply src/classic/clvm_tools/stages/stage_2/module.rs:422]: Also added a more complex test.  Please tell me what you think of the above.

File: src/classic/clvm_tools/stages/stage_2/module.rs (hunk 3 above)
Comment (prozacchiwawa) [reply src/classic/clvm_tools/stages/stage_2/module.rs:431]: no because that'd consume the collection, and each pass only takes some of the items.

File: src/classic/clvm_tools/stages/stage_2/module.rs (hunk 4 above)
Comment (prozacchiwawa) [reply src/classic/clvm_tools/stages/stage_2/module.rs]: ok nice error bubbling.

File: src/classic/clvm_tools/stages/stage_2/module.rs [hunk 6]
```diff
@@ -489,7 +491,7 @@ fn compile_mod_stage_1(
                     }
                 }
 
-                let uncompiled_main = alist[alist.len() - 1];
+                let uncompiled_main = nonempty_last(allocator.null(), &alist)?;
```
Comment (arvidn) [inline src/classic/clvm_tools/stages/stage_2/module.rs:494]: you could also do: `alist.last().unwrap()`. I believe that would have the same semantics as the subscript, it panics if the vector is empty.

If you want a message, you could do: `alist.last().expect("alist is not supposed to be empty")`

File: src/tests/classic/run.rs [hunk 7]
```diff
@@ -308,6 +308,67 @@ fn test_treehash_constant_embedded_classic() {
...
+              (defconst A 9)
+              (defconst B (* A A))
+              H
+              )
+        "}
+        .to_string(),
+    ])
+    .trim()
+    .to_string();
+    let result_hash = do_basic_brun(&vec!["brun".to_string(), result_text, "()".to_string()])
```
Comment (arvidn) [inline src/tests/classic/run.rs:340]: it's not clear to me why you execute the resulting clvm here. Don't you want to ensure that the clvm code has a specific shape? e.g. you want to make sure all the constants are folded down to a single atom. Running the code here makes the test pass even if the constants are not folded.

I suppose one, indirect, way you could test it would be to assert that the cost is low enough to indicate it was just a constant. But it would seem more straight-forward to just check the resulting clvm code from "run".

File: src/classic/clvm_tools/stages/stage_2/module.rs [hunk 8]
```diff
@@ -353,10 +354,11 @@ fn parse_mod_sexp(
...
                 Ok(())
             } else if op == "defconst".as_bytes() {
-                let r_of_declaration = rest(allocator, declaration_sexp)?;
-                let rr_of_declaration = rest(allocator, r_of_declaration)?;
-                let frr_of_declaration = first(allocator, rr_of_declaration)?;
-                delayed_constants.insert(name, frr_of_declaration);
+                // Use a new type-based match language.
+                let Rest::Here(Rest::Here(First::Here(definition))) =
+                    Rest::Here(Rest::Here(First::Here(ThisNode::Here))).
+                    select_nodes(allocator, declaration_sexp)?;
```
Comment (arvidn) [inline src/classic/clvm_tools/stages/stage_2/module.rs:360]: it's not clear that this is any easier to read. but it's not a big deal.

File: src/classic/clvm_tools/stages/stage_2/module.rs (hunk 6 above)
Comment (prozacchiwawa) [reply src/classic/clvm_tools/stages/stage_2/module.rs:494]: It's bad if this code panics (and the array index did, but the list had already been checked) because it's run from python.  When changing it i want it to be more conservative than how it was left.

File: src/tests/classic/run.rs (hunk 7 above)
Comment (prozacchiwawa) [reply src/tests/classic/run.rs:340]: ya i'll put in a test for shape as well.

File: src/classic/clvm_tools/stages/stage_2/module.rs (hunk 8 above)
Comment (prozacchiwawa) [reply src/classic/clvm_tools/stages/stage_2/module.rs:360]: Was working up this match language for more general use in more complex optimization and compiling (and eventual refactoring) anyway.

**Review by arvidn — APPROVED** [review]

File: src/classic/clvm_tools/stages/stage_2/module.rs (hunk 8 above)
Comment (emlowe) [reply src/classic/clvm_tools/stages/stage_2/module.rs:360]: Resolving conversation (merge step)
